hazelcast/src/main/java/com/hazelcast/util/MPSCQueue.java

```diff
@@ -0,0 +1,284 @@
+        }
+        this.owningThread = owningThread;
+        this.array = new Object[INITIAL_ARRAY_SIZE];
+        this.spin = spin;
+    }
+
+    @Override
+    public void clear() {
+        head.set(null);
+    }
+
+    @Override
+    public boolean offer(E value) {
+        if (value == null) {
+            throw new IllegalArgumentException("value can't be null");
+        }
+
+        AtomicReference<Node> head = this.head;
+        Node newHead = new Node();
+        newHead.value = value;
+
+        for (; ; ) {
+            Node oldHead = head.get();
+            if (oldHead == null || oldHead == BLOCKED) {
+                newHead.next = null;
+                newHead.size = 1;
+            } else {
+                newHead.next = oldHead;
+                newHead.size = oldHead.size + 1;
+            }
+
+            if (!head.compareAndSet(oldHead, newHead)) {
+                continue;
+            }
+
+            if (oldHead == BLOCKED) {
+                unpark(owningThread);
+            }
+
+            return true;
+        }
+    }
+
+    @Override
+    public E take() throws InterruptedException {
+        if (spin) {
+            for (; ; ) {
+                E item = poll();
+                if (item != null) {
+                    return item;
+                }
+            }
+        }
+
+        E item = next();
+        if (item != null) {
+            return item;
+        }
+
+        takeAll();
+        assert index == 0;
+        assert array[index] != null;
+
+        return next();
+    }
+
+    @Override
+    public E poll() {
+        E item = next();
+
+        if (item != null) {
+            return item;
+        }
+
+        if (!pollAll()) {
+            return null;
+        }
+
+        return next();
+    }
+
+
+    private E next() {
+        if (index == -1) {
+            return null;
+        }
+
+        if (index == array.length) {
+            index = -1;
+            return null;
+        }
+
+        E item = (E) array[index];
+        if (item == null) {
+            index = -1;
+            return null;
+        }
+        array[index] = null;
+        index++;
+        return item;
+    }
+
+    public void takeAll() throws InterruptedException {
+        AtomicReference<Node> head = this.head;
+        for (; ; ) {
+            Node currentHead = head.get();
+
+            if (currentHead == null) {
+
+                // there is nothing to be take, so lets block.
+                if (!head.compareAndSet(null, BLOCKED)) {
+                    continue;
+                }
+
+                park();
+            } else if (currentHead == BLOCKED) {
+                park();
+            } else {
+                if (!head.compareAndSet(currentHead, null)) {
+                    continue;
+                }
+
+                initArray(currentHead);
+                break;
+            }
+        }
+
+        if (owningThread.isInterrupted()) {
+            head.compareAndSet(BLOCKED, null);
```

Comment:
how this can ever succeed? if the head is `BLOCKED` then you will never exit the main loop, won't you?